File: example/xx_three/two/light_point.js
// light_point.js

var viewer = two.ui.scene_viewer(panel);
var controls = two.ui.trackball_controller(viewer);
//controls.dynamicDampingFactor = 0.15;
//controls.keys = [ 65, 83, 68 ];

var scene = viewer.scene;

if (init) {
    var camera = viewer.camera;
    camera.fov = 50.0; camera.near = 1.0; camera.far = 300.0;
    camera.eye = new two.vec3(0.0, 15.0, 150.0);

    var bg = two.to_linear(two.rgb(0x040306));
    viewer.viewport.to_gamma = true;
    viewer.viewport.clear_colour = bg;

    var env = scene.env;
    env.fog.enabled = true;
    env.fog.colour = bg;
    env.fog.depth_begin = 10.0;
    env.fog.depth_end = 300.0;

    // GROUND

    var groundmat = app.gfx.materials.create('ground'); var m = groundmat;
        m.program = app.gfx.programs.fetch('pbr/phong');
        m.base.uv0_scale = new two.vec2(20.0, 10.0);
        m.phong.diffuse.texture = app.gfx.textures.file('disturb.jpg');

    var zeroq = new two.quat(new two.vec3(0.0));
    
    var model = app.gfx.shape(new two.Rect(new two.vec2(0.0), new two.vec2(800.0, 400.0)));
    var node = scene.nodes().add(new two.Node3(new two.vec3(0, -5, 0), zeroq));
    scene.items().add(new two.Item(node, model, 0, groundmat));

    // OBJECTS
    
    //var material = two.gfx.pbr_material(app.gfx, 'lightpoint', two.rgb(0xffffff), 0.5, 1.0);

    var material = app.gfx.materials.create('lightpoint'); m = material;
        m.program = app.gfx.programs.fetch('pbr/three');
        m.pbr.albedo.value = two.rgb(0xffffff);
        m.pbr.metallic.value = 1.0;
        m.pbr.roughness.value = 0.5;
        
    var symbol = new two.Symbol(new two.Colour(1.0)); symbol.subdiv = new two.uvec2(8, 16);
    //var torus = app.gfx.shape(new two.Torus(1.0, 0.1));
    var torus = app.gfx.shape(new two.Torus(1.5, 0.4), symbol);
    
    for (var i = 0; i < 5000; i ++) {

        var x = 400.0 * (0.5 - Math.random());
        var y = 50.0 * (0.5 - Math.random()) + 25;
        var z = 200.0 * (0.5 - Math.random());

        var a = 3.14 * (0.5 - Math.random());
        var b = 3.14 * (0.5 - Math.random());

        //var n = scene.nodes().add(new two.Node3(new two.vec3(x, y, z), new two.quat(a, b, 0, 1)));
        var n = scene.nodes().add(new two.Node3(new two.vec3(x, y, z), new two.quat(new two.vec3(a, b, 0))));
        var it = scene.items().add(new two.Item(n, torus, 0, material));
    }

    // LIGHTS

    var intensity = 2.5;
    var distance = 100;
    var decay = 2.0;

    var colours = [0xff0040, 0x0040ff, 0x80ff80, 0xffaa00, 0x00ffaa, 0xff1100];

    var sphere = app.gfx.shape(new two.Sphere(0.25));

    this.lights = [];

    for(var i = 0; i < 6; ++i)
    {
        var c = two.rgb(colours[i]);
        var m = two.gfx.solid_material(app.gfx, 'light' + i.toString(), c);
        var n = scene.nodes().add(new two.Node3());
        var l = scene.lights().add(new two.Light(n, two.LightType.Point, false, c, intensity, distance));
        l.attenuation = decay;
        var it = scene.items().add(new two.Item(n, sphere, 0, m));
        
        this.lights.push(n);
    }

    var dn = scene.nodes().add(new two.Node3(new two.vec3(0.0), two.look_dir(new two.vec3(-0.5, -1, 0))));
    scene.lights().add(new two.Light(dn, two.LightType.Direct, false, two.rgb(0xffffff), 0.05));
}

var coef0 = [0.7, 0.3, 0.7, 0.3, 0.3, 0.7];
var coef1 = [0.3, 0.7, 0.5, 0.5, 0.5, 0.5];

var time = app.gfx.time * 0.2;

var d = 150;

for(var i = 0; i < 6; ++i)
{
    var pos = new two.vec3(Math.sin(time * coef0[i]) * d, 0.0, Math.cos(time * coef1[i]) * d);
    this.lights[i].apply(pos);
}

File: example/xx_three/light_point.cpp
#include <xx_three/xx_three.h>
#include <gfx-pbr/Api.h>

using namespace two;

EX(xx_light_point)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
#endif

#if UI
	TrackballController& controls = ui::trackball_controller(viewer);
	controls.m_dynamicDampingFactor = 0.15f;
#endif
	
	//controls.keys = [65, 83, 68];

	static vector<Node3*> lights;

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 50.f; camera.m_near = 1.f; camera.m_far = 300.f;
		camera.m_eye = vec3(0.f, 15.f, 150.f);

		const Colour bg = to_linear(rgb(0x040306));
		viewer.m_viewport.m_to_gamma = true;
		viewer.m_viewport.m_clear_colour = bg;

		Zone& env = scene.m_env;
		env.m_fog = { true, 1.f, bg, true, 10.f, 300.f };

		Material& groundmat = app.m_gfx.materials().create("ground", [&](Material& m) {
			m.m_program = &app.m_gfx.programs().fetch("pbr/phong");
			m.m_base.m_uv0_scale = vec2(20.f, 10.f);
			m.m_phong.m_diffuse = app.m_gfx.textures().file("disturb.jpg");
		});

		// GROUND

		Model& model = app.m_gfx.shape(Rect(vec2(0.f), vec2(800.f, 400.f)));
		Node3& node = gfx::nodes(scene).add(Node3(vec3(0.f, -5.f, 0.f), quat(vec3(0.f))));
		gfx::items(scene).add(Item(node, model, 0, &groundmat));

		// OBJECTS

		Material& material = app.m_gfx.materials().create("lightpoint", [&](Material& m) {
			m.m_program = &app.m_gfx.programs().fetch("pbr/three");
			m.m_pbr.m_albedo = rgb(0xffffff);
			m.m_pbr.m_metallic = 1.f;
			m.m_pbr.m_roughness = 0.5f;
		});

		Symbol symbol; symbol.m_subdiv = uvec2(8, 16);
		Model& torus = app.m_gfx.shape(Torus(1.5f, 0.4f), symbol);
		//Model& torus = app.m_gfx.shape(Torus(1.f, 0.1f));

		for(int i = 0; i < 5000; i++) {

			float x = 400.f * (0.5f - randf());
			float y = 50.f * (0.5f - randf()) + 25;
			float z = 200.f * (0.5f - randf());

			float a = 3.14f * (0.5f - randf());
			float b = 3.14f * (0.5f - randf());

			uint32_t flags = ItemFlag::Default;// | ItemFlag::NoCull;
			//Node3& n = gfx::nodes(scene).add(Node3(vec3(x, y, z), quat(a, b, 0, 1)));
			Node3& n = gfx::nodes(scene).add(Node3(vec3(x, y, z), quat(vec3(a, b, 0.f))));
			Item& it = gfx::items(scene).add(Item(n, torus, flags, &material));
			UNUSED(it);
		}

		// LIGHTS

		constexpr float intensity = 2.5f;
		constexpr float distance = 100.f;
		constexpr float decay = 2.0f;

		constexpr uint32_t colours[] = { 0xff0040, 0x0040ff, 0x80ff80, 0xffaa00, 0x00ffaa, 0xff1100 };

		Sphere sphere = Sphere(0.25f);
		Model& sphere_model = app.m_gfx.shape(sphere);

		lights.clear();
		for(int i = 0; i < 6; ++i)
		{
			const Colour c = rgb(colours[i]);
			Material& m = gfx::solid_material(app.m_gfx, ("light" + to_string(i)).c_str(), c);
			Node3& n = gfx::nodes(scene).add(Node3());
			Light& l = gfx::lights(scene).add(Light(n, LightType::Point, false, c, intensity, distance));
			l.m_attenuation = decay;
			Item& it = gfx::items(scene).add(Item(n, sphere_model, ItemFlag::Default, &m));
			UNUSED(it);

			lights.push_back(&n);
		}

		Node3& dn = gfx::nodes(scene).add(Node3(vec3(0.f), look_dir(-vec3(0.5f, 1.f, 0.f))));
		gfx::lights(scene).add(Light(dn, LightType::Direct, false, rgb(0xffffff), 0.05f));
	}

	constexpr float coef0[] = { 0.7f, 0.3f, 0.7f, 0.3f, 0.3f, 0.7f };
	constexpr float coef1[] = { 0.3f, 0.7f, 0.5f, 0.5f, 0.5f, 0.5f };

	constexpr float d = 150.f;

	const float time = app.m_gfx.m_time * 0.2f;

	for(int i = 0; i < 6; ++i)
	{
		vec2 pos = vec2(sin(time * coef0[i]) * d, cos(time * coef1[i]) * d);
		lights[i]->apply(vec3(pos.x, 0.f, pos.y));
	}
}
